File: Python/Algorithms/SearchingAlgorithms/ExponentialSearch.py
# Implementation of Exponential Search

# Python 3 Code will expand range of searching exponentially
# until required element is included and then
# perform Binary Search within that range

def ExponentialSearch(array, size, x):
    if(array[0]) == x:  # Checking for required element in first position
        return 0
    i = 1
    while (i < size) and (array[i] <= x):  # Expanding range
        i = i * 2
    return BinarySearch(array, i / 2, min(i, size), x)


def BinarySearch(array, left, right, x):  # Basic Binary Search Algorithm
    if (right >= left):
        mid = int(left + (right - left) / 2)
        if (array[mid] == x):
            return mid
        if (array[mid] > x):
            return BinarySearch(array, left, mid - 1, x)
        return BinarySearch(array, mid + 1, right, x)
    return -1  # Returns -1 when element found withing range by Binary Search


# Driver Code
array = [12, 20, 69, 420, 666]
size = len(array)
x = 420
result = ExponentialSearch(array, size, x)
if result == -1:
    print("Element doesn't exist in the array!")
else:
    print("Element exists within array at index position %d!" % result)

# Example Output:
#  Element exists within array at index position 3!

File: C++/Algorithms/SearchingAlgorithms/exponentialSearch.cpp
/*
finding the element ele using exponential search
*/

#include <bits/stdc++.h>
using namespace std;

int binary_search(int arr[], int, int, int);

//exponential search - returns the position of first occurrence of ele in array
int exponential_search(int arr[], int size, int ele)
{

    //If ele is present at arr[0]
    if (arr[0] == ele)
        return 0;

    //loop for finding the range
    //binary search to find the ele
    int i = 1;
    for (; i < size && arr[i] <= ele; i = i * 2){}

    //binary search to find the search ele within the array range
    return binary_search(arr, i / 2, min(i, size), ele);
}

//Recursive binary search -  returns the location of the search ele
int binary_search(int arr[], int left, int right, int ele)
{
    if (right >= left)
    {

        //middle element location in the array
        int middle = left + (right - left) / 2;

        //If the ele is present at the arr[middle] location
        if (arr[middle] == ele)
            return middle;

        //If the ele is smaller the middle element than
        //the ele is present in left subarray
        if (arr[middle] > ele)
        {
            return binary_search(arr, left, middle - 1, ele);

            //ele present in right subarray
            return binary_search(arr, middle + 1, right, ele);
        }
    }

    //If the search element is not present the array
    return -1;
}

//starting point
int main(void)
{
    //variables declaration
    int size, i, ele;
    int * arr;

    //asking for no.of elements to be added 
    cout<<"How many numbers you would like to add? "<<"\n";
    cin>>size;
    
    try
    {
        arr = new int[size];
        if(arr == nullptr){
            cout << "Error: memory could not be allocated";
            throw 505;
        }
        else{
            cout<<"Enter the numbers: "<<"\n";
            for(i=0; i < size; i++){
                cin>>arr[i];
            }
        }
        cout<<"Enter the element to be search?"<<"\n";
        cin>>ele;
        int result = exponential_search(arr, size, ele);

        //displaying the result whether the element is present or not
        (result == -1) ? printf("Element is not present in array")
                    : printf("Element is present at index %d",
                                result+1);
        
    }
    //catching the unknown error
    catch(...)
    {
        cout<<"Error occurred";
    }
    
    return 0;
}